faster type allocation lookup

vtjnash opened the pull request:
This previously implemented a binary search to find the type, but that required holding a lock and doing upwards of 16 steps of comparisons. This uses instead a hash set (nearly an IdDict). I observe a load factor of about 30% for Tuple, and much faster lookups.
On this microbenchmark test script, it shows a >2x improvement:
```julia
julia> T = Int; @btime Tuple{T, T}
before:  172.414 ns (0 allocations: 0 bytes)
after:    62.356 ns (0 allocations: 0 bytes)
```

vtjnash committed:
types: improve insertion and lookup performance

Currently, this uses an expensive tree lookup and requires a lock.
Switching this to a hash function instead of a comparison allows us
to make this much faster, especially for insertion.

vtjnash committed:
maybe make typekey_eq faster when either key should have a unique id

JeffBezanson commented:
Any measurable changes in package loading?

vtjnash commented:
It's not a noticeable percentage of the cost there, but it's nice to set it to effectively zero (fixes much of #34890—possibly because this hits the L1/L2 caches much less hard??).
on master:
```
$ time ./usr/bin/julia -e 'using DataFrames'; time ./usr/bin/julia -e 'using Plots'; time ./usr/bin/julia -e 'using ApproxFun'
real	0m1.151s
user	0m1.289s
sys	0m0.144s

real	0m4.243s
user	0m4.248s
sys	0m0.207s

real	0m18.912s
user	0m18.582s
sys	0m0.858s
```
current measurements (with many additional changes):
```
$ time ./usr/bin/julia -e 'using DataFrames'; time ./usr/bin/julia -e 'using Plots'; time ./usr/bin/julia -e 'using ApproxFun'

real	0m1.175s
user	0m1.268s
sys	0m0.156s

real	0m3.843s
user	0m3.901s
sys	0m0.199s

real	0m16.749s
user	0m16.317s
sys	0m0.995s
```

JeffBezanson commented on src/jltypes.c:
```diff
@@ -574,67 +581,36 @@ static int is_typekey_ordered(jl_value_t **key, size_t n)
+    for (j = 0; j < n; j++) {
+        jl_value_t *kj = key[j];
+        jl_datatype_t *dk = (jl_datatype_t*)(jl_is_unionall(kj) ? jl_unwrap_unionall(kj) : kj);
+        if (!jl_is_datatype(kj))
```
Did you mean `dk` instead of `kj` here?

JeffBezanson commented on src/jltypes.c:
```diff
@@ -748,54 +724,126 @@ static int is_cacheable(jl_datatype_t *type)
+        assert(jl_is_datatype(type));
+        jl_value_t **key = jl_svec_data(type->parameters);
+        int n = jl_svec_len(type->parameters);
+        int ord = is_typekey_ordered(key, n);
```
Rename this `is_typekey_hashable`?

vtjnash commented:
I'm going to recreate this as a different PR that goes further and incorporates other work.

vtjnash committed:
use datatype->hash for object_id for concrete types

these may get uniqued during construction, so it is not necessary to check their content

vtjnash committed:
iddict: ensure compiler only loads tab[index] once

I believe this should make it thread-safe to do lookup in the presence of concurrent writers.

vtjnash committed:
types: redesign code to eliminate datatype uid

Use a normal IdDict instead of a fancy uid-indexed, and optimize
object_id to use a cached hash field for them.

vtjnash committed:
typecache: fix bug in hash collision handling

In the precompile test, we intentionally create objects whose hashes
here should collide. Previously, we'd take a random branch here, which
could fail in as much as 25% of cases.

vtjnash committed:
types: make cache lookup thread-safe

Since this is not a binary-tree anymore, we can also stop holding a lock
around the operation, which is a significant optimization opportunity.

vtjnash commented:
Today's measurements:
master
```
$ time ./usr/bin/julia -e 'using DataFrames'; time ./usr/bin/julia -e 'using Plots'; time ./usr/bin/julia -e 'using ApproxFun'
real	0m1.122s
user	0m1.241s
sys	0m0.153s

real	0m3.811s
user	0m3.863s
sys	0m0.196s

real	0m20.926s
user	0m20.581s
sys	0m0.814s

time ./julia -e 'using Plots; @time display(@time plot(1:10))'
  4.062018 seconds (9.54 M allocations: 473.947 MiB, 5.78% gc time)
  9.953003 seconds (17.81 M allocations: 876.303 MiB, 3.06% gc time)

real	0m13.985s
user	0m13.827s
sys	0m0.395s

julia> T = Int; @btime Tuple{T, T}
  109.337 ns (0 allocations: 0 bytes)
```
PR
```

real	0m1.056s
user	0m1.196s
sys	0m0.152s

real	0m3.600s
user	0m3.677s
sys	0m0.206s

real	0m16.996s
user	0m16.774s
sys	0m0.772s

time ./julia -e 'using Plots; @time display(@time plot(1:10))'
  3.862645 seconds (9.55 M allocations: 474.279 MiB, 7.42% gc time)
  9.783306 seconds (17.81 M allocations: 876.683 MiB, 3.55% gc time)

real	0m13.676s
user	0m13.499s
sys	0m0.419s

julia> T = Int; @btime Tuple{T, T}
  41.030 ns (0 allocations: 0 bytes)
```

StefanKarpinski commented:
Those are substantial savings. It's kind of weird that with all the other things getting faster the literal time-to-first-plot doesn't get much better, but solving that isn't the _only_ concern, so 🎉

JeffBezanson commented:
Ick, doctests --- this seems awfully implementation-detaily to have doctests for the output; maybe we should edit that section a bit.

JeffBezanson commented on src/builtins.c:
```diff
@@ -324,11 +332,10 @@ JL_DLLEXPORT uintptr_t jl_object_id_(jl_value_t *tv, jl_value_t *v) JL_NOTSAFEPO
-        return bitmix(~dtv->name->hash, hash_svec(dtv->parameters));
+        uintptr_t h = ~dtv->name->hash;
+        if (dtv->isconcretetype)
+            return bitmix(h, dtv->hash);
```
Why isn't `dtv->hash` sufficient here?

vtjnash replied:
dtv->hash was only the contents of the parameters, so we need to add in the type itself now, but I could change that

JeffBezanson replied:
Yes, it seems to make sense to me to have the name's hash already mixed in. Any downside?

vtjnash commented:
Yeah, I need to take that doctest out (the hash is a bitmix that includes the module build_id random number). Additionally / alternatively, I've considered making small dictionaries use a linear lookup.

JeffBezanson commented on src/iddict.c:
```diff
@@ -151,22 +148,24 @@ jl_array_t *jl_eqtable_put(jl_array_t *h, jl_value_t *key, jl_value_t *val, int
     return h;
 }
 
+// Note: lookup in the IdDict it permitted concurrently, if you avoid deletions,
```
```suggestion
// Note: lookup in the IdDict is permitted concurrently, if you avoid deletions,
```
